src/execution/operator/join/physical_range_join.cpp

```diff
@@ -0,0 +1,357 @@
+#include "duckdb/common/fast_mem.hpp"
+#include "duckdb/common/operator/comparison_operators.hpp"
+#include "duckdb/common/row_operations/row_operations.hpp"
+#include "duckdb/common/sort/comparators.hpp"
+#include "duckdb/common/sort/sort.hpp"
+#include "duckdb/common/vector_operations/vector_operations.hpp"
+#include "duckdb/execution/expression_executor.hpp"
+#include "duckdb/main/client_context.hpp"
+#include "duckdb/parallel/event.hpp"
+#include "duckdb/parallel/thread_context.hpp"
+
+#include <thread>
+
+namespace duckdb {
+
+PhysicalRangeJoin::LocalSortedTable::LocalSortedTable(const PhysicalRangeJoin &op, const idx_t child)
+    : op(op), has_null(0), count(0) {
+	// Initialize order clause expression executor and key DataChunk
+	vector<LogicalType> types;
+	for (const auto &cond : op.conditions) {
+		const auto &expr = child ? cond.right : cond.left;
+		executor.AddExpression(*expr);
+
+		types.push_back(expr->return_type);
+	}
+	keys.Initialize(types);
+}
+
+void PhysicalRangeJoin::LocalSortedTable::Sink(DataChunk &input, GlobalSortState &global_sort_state) {
+	// Initialize local state (if necessary)
+	if (!local_sort_state.initialized) {
+		local_sort_state.Initialize(global_sort_state, global_sort_state.buffer_manager);
+	}
+
+	// Obtain sorting columns
+	keys.Reset();
+	executor.Execute(input, keys);
+
+	// Count the NULLs so we can exclude them later
+	has_null += MergeNulls(op.conditions);
+	count += keys.size();
+
+	//	Only sort the primary key
+	DataChunk join_head;
+	join_head.data.emplace_back(Vector(keys.data[0]));
+	join_head.SetCardinality(keys.size());
+
+	// Sink the data into the local sort state
+	local_sort_state.SinkChunk(join_head, input);
+}
+
+PhysicalRangeJoin::GlobalSortedTable::GlobalSortedTable(ClientContext &context, const vector<BoundOrderByNode> &orders,
+                                                        RowLayout &payload_layout)
+    : global_sort_state(BufferManager::GetBufferManager(context), orders, payload_layout), has_null(0), count(0),
+      memory_per_thread(0) {
+	D_ASSERT(orders.size() == 1);
+
+	// Set external (can be force with the PRAGMA)
+	auto &config = ClientConfig::GetConfig(context);
+	global_sort_state.external = config.force_external;
+	// Memory usage per thread should scale with max mem / num threads
+	// We take 1/4th of this, to be conservative
+	idx_t max_memory = global_sort_state.buffer_manager.GetMaxMemory();
+	idx_t num_threads = TaskScheduler::GetScheduler(context).NumberOfThreads();
+	memory_per_thread = (max_memory / num_threads) / 4;
+}
+
+void PhysicalRangeJoin::GlobalSortedTable::Combine(LocalSortedTable &ltable) {
+	global_sort_state.AddLocalState(ltable.local_sort_state);
+	has_null += ltable.has_null;
+	count += ltable.count;
+}
+
+void PhysicalRangeJoin::GlobalSortedTable::IntializeMatches() {
+	found_match = unique_ptr<bool[]>(new bool[Count()]);
+	memset(found_match.get(), 0, sizeof(bool) * Count());
+}
+
+void PhysicalRangeJoin::GlobalSortedTable::Print() {
```

Comment:
Could this Print function be moved into the GlobalSortState? It seems like it's more generally applicable than just the PhysicalRangeJoin